packages/runtime/src/parser/parser.ts

```diff
@@ -241,33 +256,84 @@ export function opcodeFromSentence (words: string[], counter: number, interprete
   // Handle Label
   if (opCode.endsWith(':')) {
     assertLen(words.length, 0, counter);
-    if (opCodeMap[opCode.slice(0, opCode.length - 1)] !== undefined) {
-      throw new TealError(ERRORS.TEAL.INVALID_LABEL, { line: counter });
+    if (opCodeMap[tealVersion][opCode.slice(0, opCode.length - 1)] !== undefined) {
+      throw new TealError(ERRORS.TEAL.INVALID_LABEL, { line: counter }); // eg. `int:` is invalid label as `int` is an opcode
     }
     return new Label([opCode], counter);
   }
 
-  if (opCodeMap[opCode] === undefined) {
-    throw new TealError(ERRORS.TEAL.UNKOWN_OPCODE, { opcode: opCode, line: counter });
+  if (opCodeMap[tealVersion][opCode] === undefined) {
+    throw new TealError(ERRORS.TEAL.UNKOWN_OPCODE, { opcode: opCode, version: tealVersion, line: counter });
   }
 
+  // increment gas of TEAL code
+  // select from OpGasCost map if cost is different than 1
+  if (opCode !== '#pragma') { interpreter.gas += OpGasCost[tealVersion][opCode] ?? 1; }
+
   if (interpreterReqList.has(opCode)) {
-    return new opCodeMap[opCode](words, counter, interpreter);
+    return new opCodeMap[tealVersion][opCode](words, counter, interpreter);
+  }
+  return new opCodeMap[tealVersion][opCode](words, counter);
+}
+
+// verify max cost of TEAL code is within consensus parameters
+function assertMaxCost (gas: number, mode: ExecutionMode): void {
+  if (mode === ExecutionMode.STATELESS) {
+    // check max cost (for stateless)
+    if (gas > LogicSigMaxCost) {
+      throw new TealError(ERRORS.TEAL.INVALID_MAX_COST, {
+        cost: gas,
+        maxcost: LogicSigMaxCost,
+        mode: 'Stateless'
+      });
+    }
+  } else {
+    if (gas > MaxAppProgramCost) {
+      // check max cost (for stateful)
+      throw new TealError(ERRORS.TEAL.INVALID_MAX_COST, {
+        cost: gas,
+        maxcost: MaxAppProgramCost,
+        mode: 'Stateful'
+      });
+    }
+  }
+}
+
+// verify max length of TEAL code is within consensus parameters
+function assertMaxLen (len: number, mode: ExecutionMode): void {
```

Comment:
does this `len` includes blank lines or just the opcodes ?

Reply:
It's actually the length of compiled teal code + arguments (for stateless) and 1kb for stateful execution mode. Not sure about how to calculate this, as we're not using the TEAL compiler.

Reply:
Added a new task to think about it: [https://www.pivotaltracker.com/story/show/176623588](https://www.pivotaltracker.com/story/show/176623588)